File: crabzilla-example1/crabzilla-example1-service/src/main/java/io/github/crabzilla/example1/Example1EventProjector.java
package io.github.crabzilla.example1;

import io.github.crabzilla.example1.readmodel.CustomerSummary;
import io.github.crabzilla.example1.util.AbstractExample1EventProjector;
import lombok.extern.slf4j.Slf4j;
import org.jdbi.v3.core.Jdbi;

import static io.github.crabzilla.example1.customer.CustomerData.*;

// tag::projector[]
@Slf4j
public class Example1EventProjector extends AbstractExample1EventProjector<CustomerSummaryDao> {

  protected Example1EventProjector(String eventsChannelId,
                                   Class<CustomerSummaryDao> customerSummaryDaoClass,
                                   Jdbi jdbi) {
    super(eventsChannelId, customerSummaryDaoClass, jdbi);
  }

  public void handle(CustomerSummaryDao dao, String targetId, CustomerCreated event) {
    dao.insert(new CustomerSummary(targetId, event.getName(), false));
  }

  public void handle(CustomerSummaryDao dao, String targetId, CustomerActivated event) {
    dao.updateStatus(targetId, true);
  }

  public void handle(CustomerSummaryDao dao, String targetId, CustomerDeactivated event) {
    dao.updateStatus(targetId, false);
  }
}
// end::projector[]

File: crabzilla-pgc/src/test/java/io/github/crabzilla/pgc/example1/Example1EventProjector.kt
package io.github.crabzilla.pgc.example1

import io.github.crabzilla.DomainEvent
import io.github.crabzilla.example1.CustomerActivated
import io.github.crabzilla.example1.CustomerCreated
import io.github.crabzilla.example1.CustomerDeactivated
import io.github.crabzilla.pgc.PgcEventProjector
import io.github.crabzilla.pgc.runPreparedQuery
import io.reactiverse.pgclient.PgTransaction
import io.reactiverse.pgclient.Tuple
import io.vertx.core.Future
import org.slf4j.LoggerFactory

class Example1EventProjector : PgcEventProjector {

  private val log = LoggerFactory.getLogger(Example1EventProjector::class.java.name)

  override fun handle(pgTx: PgTransaction, targetId: Int, event: DomainEvent): Future<Void> {

    val future = Future.future<Void>()

    log.info("event {} ", event)

    when (event) {
      is CustomerCreated -> {
        val query = "INSERT INTO customer_summary (id, name, is_active) VALUES ($1, $2, $3)"
        val tuple = Tuple.of(targetId, event.name, false)
        pgTx.runPreparedQuery(query, tuple, future)
      }
      is CustomerActivated -> {
        val query = "UPDATE customer_summary SET is_active = true WHERE id = $1"
        val tuple = Tuple.of(targetId)
        pgTx.runPreparedQuery(query, tuple, future)
      }
      is CustomerDeactivated -> {
        val query = "UPDATE customer_summary SET is_active = false WHERE id = $1"
        val tuple = Tuple.of(targetId)
        pgTx.runPreparedQuery(query, tuple, future)
      }
      else -> {
        val error = "${event.javaClass.simpleName} does not have any event projector handler"
        log.info(error)
        future.complete()
      }
    }

    log.info("finished event {} ", event)

    return future

  }

}
